libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,331 @@
+            _state.last_time_micros = tnow;
+            _state.healthy = true;
+        }
+    } else {
+        _has_cell_voltages = false;
+    }
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
+{
+    uint16_t data;
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+
+        if((fabs(temp - avg(temperatures)) < 100) || (temperatures.size() == 0 && temp < 1000)) {
+            temperatures.push_back(temp);
+            if(temperatures.size() > 10) {
+                temperatures.erase(temperatures.begin());
+            }
+        }
+    }
+
+    _state.temperature = avg(temperatures);
+    // gcs().send_text(MAV_SEVERITY_WARNING, "temperature is %.0f", _state.temperature);
+    _state.temperature_time = AP_HAL::millis();
+
+    return true;
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_remaining_capacity() {
+    int32_t capacity = _params._pack_capacity;
+
+    if (capacity > 0) {
+        uint16_t data;
+        if (read_word(BATTMONITOR_SMBUS_REMAINING_CAPACITY, data)) {
+            float consumed = (capacity - data);
+            if((consumed > 0) && (consumed < capacity)) {
+                consumed_mahs.push_back(consumed);
+
+                if(consumed_mahs.size() > 10) {
+                    consumed_mahs.erase(consumed_mahs.begin());
+                }
+
+                _state.consumed_mah = avg(consumed_mahs);
+                // _state.consumed_mah = consumed;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+
+// read_block - returns number of characters read if successful, zero if unsuccessful
+uint8_t AP_BattMonitor_SMBus_SUI::read_block(uint8_t reg, uint8_t* data, uint8_t max_len, bool append_zero) const
+{
+    uint8_t buff[max_len+2];    // buffer to hold results (2 extra byte returned holding length and PEC)
+
+    // read bytes
+    if (!_dev->read_registers(reg, buff, sizeof(buff))) {
+        return 0;
+    }
+
+    // get length
+    uint8_t bufflen = buff[0];
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > max_len) {
+        return 0;
+    }
+
+    // check PEC
+    uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
+    if (pec != buff[bufflen+1]) {
+        return 0;
+    }
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // optionally add zero to end
+    if (append_zero) {
+        data[bufflen] = '\0';
+    }
+
+    // return success
+    return bufflen;
+}
+
+// read_bare_block - returns number of characters read if successful, zero if unsuccessful
+uint8_t AP_BattMonitor_SMBus_SUI::read_block_bare(uint8_t reg, uint8_t* data, uint8_t max_len, bool append_zero) const
+{
+    // read bytes
+    if (!_dev->read_registers(reg, data, max_len)) {
+        return 0;
+    }
+
+    // return success
+    return max_len;
+}
+
+AP_BattMonitor_WithCellCount::AP_BattMonitor_WithCellCount(AP_BattMonitor &mon,
```

Comment:
What is this wrapper buying you? Just seems to make it more confusing, why not just instantiate the SUI one directly?

Reply:
removed